opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -2539,6 +2555,127 @@ namespace Opm {
 
 
 
+
+
+    template<typename TypeTag>
+    void
+    BlackoilWellModel<TypeTag>::
+    runWellPIScaling(const int timeStepIdx, DeferredLogger& local_deferredLogger)
+    {
+        if (this->last_run_wellpi_.has_value() && (*this->last_run_wellpi_ == timeStepIdx)) {
+            // We've already run WELPI scaling for this report step.  Most
+            // common for the very first report step.  Don't redo WELPI scaling.
+            return;
+        }
+
+        auto hasWellPIEvent = [this, timeStepIdx](const int well_index) -> bool
+        {
+            return this->schedule()
+                .hasWellGroupEvent(this->wells_ecl_[well_index].name(),
+                                   ScheduleEvents::Events::WELL_PRODUCTIVITY_INDEX,
+                                   timeStepIdx);
+        };
+
+        auto getWellPI = [this](const int well_index) -> double
+        {
+            const auto& pu = this->phase_usage_;
+            const auto  np = this->numPhases();
+            const auto* pi = &this->well_state_.productivityIndex()[np*well_index + 0];
+
+            const auto preferred = this->wells_ecl_[well_index].getPreferredPhase();
+            switch (preferred) { // Should really have LIQUID = OIL + WATER here too...
+            case Phase::WATER:
+                return pu.phase_used[BlackoilPhases::PhaseIndex::Aqua]
+                    ? pi[pu.phase_pos[BlackoilPhases::PhaseIndex::Aqua]]
+                    : 0.0;
+
+            case Phase::OIL:
+                return pu.phase_used[BlackoilPhases::PhaseIndex::Liquid]
+                    ? pi[pu.phase_pos[BlackoilPhases::PhaseIndex::Liquid]]
+                    : 0.0;
+
+            case Phase::GAS:
+                return pu.phase_used[BlackoilPhases::PhaseIndex::Vapour]
+                    ? pi[pu.phase_pos[BlackoilPhases::PhaseIndex::Vapour]]
+                    : 0.0;
+
+            default:
+                throw std::invalid_argument {
```

Comment:
I am not updated with this. Are we supposed to use `DeferredLogger`  facility for this? @atgeirr ?

Reply:
I do not think we must use it here, but the calling code should ideally try-catch, communicate failure etc. However it is very annoying to use such elaborate things for a simple "missing enum value" type check, perhaps we should rather return an `std::optional` here if it goes wrong instead?

Reply:
> perhaps we should rather return an `std::optional` here if it goes wrong instead?

Then the question becomes what to do if there is no defined value in the `optional`.  In that case, there really is no well-defined way to continue the simulation since the input model requests something that we cannot support.